Control/setting changes

mikechristie opened the pull request:
This patchset converts gwcli to the ceph-iscsi-config changes in PR:

https://github.com/ceph/ceph-iscsi-config/pull/77

which began to make the controls/setting code easier to use across multiple apps. It also begins to make it easier to add new arguments to the reconfigure and create commands. We now only need to update the help info.

dillaman commented on gwcli/gateway.py:
```diff
@@ -414,7 +414,7 @@ def _set_controls(self, controls):
         return "Gateways: {}".format(len(self.gateway_group.children)), None
 
-    def ui_command_reconfigure(self, attribute, value):
+    def ui_command_reconfigure(self, attribute=None, value=None, **kwargs):
```
Should ```attribute``` and ```value``` be removed? What happens if someone uses ```reconfigure attribute=cmdsn_depth value=128```?

mikechristie replied:
I left it for compat support in case people had scripts that were interacting with the command already:

If you do

reconfigure attribute=cmdsn_depth value=12

then it still works like before. I just detect that and stick it in the controls dict that gets passed around.

If you mix like this:

reconfigure attribute=cmdsn_depth value=12 cmdsn_depth=13

then we would end up overwriting the second one (13) with the first (12). I do not think that issue will come up a lot.


If you mix like this:

reconfigure attribute=cmdsn_depth value=12 immediate_data=no

then we would detect that and pass both cmdsn_depth and immediate data in the controls dict that gets passed around.

Same with the below calls too.

If we want to ignore compat it is ok with me and I can drop that code. I think the chances of users having a script that calls these commands is very unlikely. I was not sure how strict we were.

dillaman replied:
Ah -- fair point. I'm fine leaving it then assuming it gets optionally added to the override dict (and prevents you from passing ```attribute= value=XYZ```).

dillaman replied:
@mikechristie Running "reconfigure  rbd.lun0 max_data_area_mb= osd_op_timeout=10" results in an error of "expected integer for max_data_area_mb". It looks like it's taking the "osd_op_timeout=10" as the param

dillaman commented on gwcli/storage.py:
```diff
@@ -140,13 +142,15 @@ def reset(self):
 
-    def ui_command_create(self, pool=None, image=None, size=None, count=1, max_data_area_mb=None):
+    def ui_command_create(self, pool=None, image=None, size=None, count=1,
+                          max_data_area_mb=None, **kwargs):
```
Nit: same comment -- can we drop the ```max_data_area_mb``` parameter?

dillaman commented on gwcli/storage.py:
```diff
@@ -376,26 +393,29 @@ def ui_command_resize(self, image_id=None, size=None):
             return
 
-    def ui_command_reconfigure(self, image_id, attribute, value):
+    def ui_command_reconfigure(self, image_id, attribute=None, value=None,
```
Nit: same comment about ```attribute``` and ```value```

dillaman commented on gwcli/storage.py:
```diff
@@ -823,23 +853,24 @@ def ui_command_resize(self, size=None):
         self.resize(size)
 
-    def ui_command_reconfigure(self, attribute, value):
+    def ui_command_reconfigure(self, attribute=None, value=None, **kwargs):
```
Nit: same

mikechristie commented:
The bug where reconfigure did not handle multiple empty attributes is fixed. I just dropped support for now, because it requires a pyparser and configshell change. You can still pass in multiple attribs during lun creation though. If you try to pass in an empty attib value for that command it will return error since it does not make sense.

I also added 2 patches to fix issues where we committed the controls to gateway.conf before we knew if they were good, so we could end up with invalid values. When you rebooted the nodes then we would fail due to trying to use the invalid values from the gateway.conf. Note that for tcmu we only recently started returning an error when a invalid value was written to the control file, so you probably did not hit a issue with older RHEL/upstream kernels.

dillaman commented:
@mikechristie I haven't dug into it yet, but with these changes, LUNs are no longer getting mapped to clients. For example, I have two clients (one w/ disks directly mapped, the other w/ disks mapped via a group). After running "/disks reconfigure rbd.lun0 max_data_area_mb 128" and looking at "targetcli ls", all the LUNs are unmapped. Restarting 'rbd-target-gw' fixes the LUN mapping for the directly mapped host, but the host-group host remains unmapped.

mikechristie commented:
I did a repush in ceph-iscsi-config to fix that issue.

I also did a repush of this cli PR to fix a bug I added where config.refresh was not getting called so gwcli would show stale info.

dillaman commented:
@mikechristie Seems to work for me now. Thanks!
